include/boost/url/impl/segments.hpp

```diff
@@ -51,161 +59,63 @@ class segments::iterator
     using pointer = void const*;
     using difference_type = std::ptrdiff_t;
     using iterator_category =
-        std::random_access_iterator_tag;
-
-    iterator() = default;
+        std::bidirectional_iterator_tag;
 
-    iterator&
-    operator++() noexcept
-    {
-        ++i_;
-        return *this;
-    }
+    iterator() noexcept = default;
 
-    iterator
-    operator++(int) noexcept
-    {
-        auto tmp = *this;
-        ++*this;
-        return tmp;
-    }
+    iterator(iterator const&) noexcept = default;
 
     iterator&
-    operator--() noexcept
-    {
-        --i_;
-        return *this;
-    }
-
-    iterator
-    operator--(int) noexcept
-    {
-        auto tmp = *this;
-        --*this;
-        return tmp;
-    }
+    operator=(iterator const&) noexcept = default;
 
-    BOOST_URL_DECL
     const_string
-    operator*() const;
-
-    friend
-    bool
-    operator==(
-        iterator a,
-        iterator b) noexcept
+    operator*() const noexcept
     {
-        BOOST_ASSERT(a.u_ == b.u_);
-        return a.u_ == b.u_ &&
-            a.i_ == b.i_;
+        return *impl_;
```

Comment:
unless the type is user-facing you shouldn't use the operators. Just use `impl_.dereference()`.